presto-main/src/main/java/com/facebook/presto/operator/scalar/FailureFunction.java

```diff
@@ -40,6 +40,24 @@ public static boolean failWithException(@SqlType(StandardTypes.JSON) Slice failu
         throw new PrestoException(StandardErrorCode.GENERIC_USER_ERROR, failureInfo.toException());
     }
 
+    // This function is only used to propagate optimization failures.
+    @Description("Decodes json to an exception and throws it with supplied errorCode")
+    @ScalarFunction(value = "fail", hidden = true)
```

Comment:
Should we give it a more specific name than `fail` like `$internal$fail` or something to avoid potential conflict with a real function.